apps/passport-server/src/apis/devconnect/devconnectPretixAPI.ts

```diff
@@ -28,9 +34,96 @@ export interface IDevconnectPretixAPI {
     orgUrl: string,
     token: string
   ): Promise<DevconnectPretixEvent[]>;
+  cancelPendingRequests(): void;
+}
+
+export interface DevconnectPretixAPIOptions {
+  tokenRequestsPerMinute?: number;
+  concurrency?: number;
+  interval?: number;
 }
 
 export class DevconnectPretixAPI implements IDevconnectPretixAPI {
+  private tokenRequestsPerMinute: number;
+  private concurrency: number;
+  private interval: number;
+  private requestQueues: Map<string, PQueue>;
+  private abortController: AbortController;
+
+  public constructor(options?: DevconnectPretixAPIOptions) {
+    // Default is 100 requests per minute
+    this.tokenRequestsPerMinute = options?.tokenRequestsPerMinute ?? 100;
+    this.interval = options?.interval ?? 60_000;
+    this.concurrency = options?.concurrency ?? 1;
+
+    this.requestQueues = new Map();
+    this.abortController = new AbortController();
+  }
+
+  private getQueue(token: string): PQueue {
+    if (!this.requestQueues.has(token)) {
+      const queue = new PQueue({
+        // @todo set these in constructor?
+        concurrency: this.concurrency,
+        intervalCap: this.tokenRequestsPerMinute,
+        interval: this.interval
+      });
+
+      this.requestQueues.set(token, queue);
+    }
+
+    return this.requestQueues.get(token) as PQueue;
+  }
+
+  private queuedFetch(
+    queue: PQueue,
+    input: RequestInfo | URL,
+    init?: RequestInit
+  ): Promise<Response> {
+    return queue.add(async () => {
+      // Create a function for doing the fetch, so we can retry it
+      const doFetch = async (): Promise<Response> => {
+        return fetch(input, {
+          signal: this.abortController.signal,
+          ...init
+        });
+      };
+
+      let result = await doFetch();
+
+      // If Pretix wants us to slow down
+      while (result.status === 429) {
+        logger(
+          `[DEVCONNECT PRETIX] Received status 429 while fetching: ${input}`
+        );
+        // Get how long to wait for
+        const replyAfter = result.headers.get("Retry-After");
+        if (replyAfter) {
+          const seconds = parseFloat(replyAfter);
+          // Wait for the specified time
+          await new Promise((f) => setTimeout(f, seconds * 1000));
+          // Try again
+          result = await doFetch();
+        } else {
+          break;
+        }
+      }
+
+      if (!result.ok) {
+        throw new Error(
+          `[PRETIX] Error fetching ${input}: ${result.status} ${result.statusText}`
+        );
+      }
+
+      return result;
+    });
+  }
+
+  public cancelPendingRequests(): void {
+    this.abortController.abort();
+    this.abortController = new AbortController();
```

Comment:
> then requests that are queued but haven't been invoked yet will not be aborted

Ugh. Yes, p-queue isn't actually queueing promises, it's queueing functions that return promises, which means that binding to `this.abortController` is later than I had thought. I can also pass the abort controller signal to `PQueue.add`, which would then cancel the pending jobs as well. I can also make the enqueued function close over the current value of `this.abortController.signal` at the point the function is created rather than executed.

The abort controller cannot be re-used.

Reply:
another thing that just came to mind is you probably want some way to cancel requests just for a single organizer, rather than for all organizers. this thought points me in the direction of wanting a separate instance of `DevconnectPretixAPI` per `OrganizerSync`, rather than having `DevconnectPretixAPI` manage the complexity of multiple organizers internally.